src/main/java/com/extendedclip/deluxemenus/listener/PlayerListener.java

```diff
@@ -148,76 +152,77 @@ public void onClick(InventoryClickEvent event) {
+              item.options().shiftRightClickRequirements().orElse(null))) {
+        return;
       }
-      this.cache.put(player.getUniqueId(), System.currentTimeMillis());
-      item.getShiftRightClickHandler().onClick(holder);
-    } else if (event.getClick() == ClickType.LEFT && item.getLeftClickHandler() != null) {
-      if (item.getLeftClickRequirements() != null) {
-        if (!item.getLeftClickRequirements().evaluate(holder)) {
-          if (item.getLeftClickRequirements().getDenyHandler() != null) {
-            item.getLeftClickRequirements().getDenyHandler().onClick(holder);
-          }
-          return;
-        }
+    }
+
+    if (event.getClick() == ClickType.LEFT) {
+      if (handleClick(player, holder, item.options().leftClickHandler().orElse(null),
+              item.options().leftClickRequirements().orElse(null))) {
+        return;
       }
-      this.cache.put(player.getUniqueId(), System.currentTimeMillis());
-      item.getLeftClickHandler().onClick(holder);
-    } else if (event.getClick() == ClickType.RIGHT && item.getRightClickHandler() != null) {
-      if (item.getRightClickRequirements() != null) {
-        if (!item.getRightClickRequirements().evaluate(holder)) {
-          if (item.getRightClickRequirements().getDenyHandler() != null) {
-            item.getRightClickRequirements().getDenyHandler().onClick(holder);
-          }
-          return;
-        }
+    }
+
+    if (event.getClick() == ClickType.RIGHT) {
+      if (handleClick(player, holder, item.options().rightClickHandler().orElse(null),
+              item.options().rightClickRequirements().orElse(null))) {
+        return;
       }
-      this.cache.put(player.getUniqueId(), System.currentTimeMillis());
-      item.getRightClickHandler().onClick(holder);
-    } else if (event.getClick() == ClickType.MIDDLE && item.getMiddleClickHandler() != null) {
-      if (item.getMiddleClickRequirements() != null) {
-        if (!item.getMiddleClickRequirements().evaluate(holder)) {
-          if (item.getMiddleClickRequirements().getDenyHandler() != null) {
-            item.getMiddleClickRequirements().getDenyHandler().onClick(holder);
-          }
-          return;
+    }
+
+    if (event.getClick() == ClickType.MIDDLE) {
+      if (handleClick(player, holder, item.options().middleClickHandler().orElse(null),
+              item.options().middleClickRequirements().orElse(null))) {
+        return;
+      }
+    }
+  }
+
+  /**
+   * Handles menu click by player
+   * @param player player who clicked
+   * @param holder menu holder
+   * @param handler click handler
+   * @param requirements click requirements
+   * @return true if click was handled successfully. will ever return false if no click handler was found
+   */
+  private boolean handleClick(final @NotNull Player player, final @NotNull MenuHolder holder,
+                              final @Nullable ClickHandler handler, final @Nullable RequirementList requirements) {

```

Comment:
I see that you use `item.options().{type}ClickHandler().orElse(null)` everywhere, perhaps replace the parameters of this method with `Optional<MenuHolder` and `Optional<ClickHandler>`? You have a null check inside this method too, which will  become an `Optional#isEmpty`

Reply:
While it is usually recommended not to use Optional as parameter types, I think you're right and this might be one of those rare cases this is acceptable. I'll change it.